### PR TITLE
fix(CurrencyPipe): should accept `null` input on currencyCode

### LINKED ISSUE #7917: bug: CurrencyPipe should accept `null` input on currencyCode
The `CurrencyPipe` should accept `null` as a valid input for the `currencyCode` argument and (in this case) behave like the `DecimalPipe`, i.e. just display the number.

If the currency (and currency code) are part of the model than there can be situations where the currency has not yet been set by the user. It would of cause be possible to work around this by using `ngIf`, but this would make the template unnecessarily complex and reduces its readability.

This behaviour would also be consistent with the fact that the number pipe accept `null` as the value and don't throw an error.

**Steps to reproduce and a minimal demo of the problem**

use the `CurrencyPipe` like this: `{{12.34 | currency:null}}`

example: http://plnkr.co/edit/VqorIUr9gahX7rcMRtzP?p=preview

**Current behavior**

The following error is thrown: `RangeError: Invalid currency code: null` and the template is not rendered and crashes.

**Expected/desired behavior**

The number should just be formatted as if the `DecimalPipe` would have been used.

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/7917

### CODE REVIEW

**Comment by IgorMinar** [comment]
I'm marking this as blocked because the arbitrarily changing currency changes the meaning of the displayed value in significant way. I've pinged @vicb and @pkozlowski-opensource to explain why this is ok.

File: modules/@angular/common/test/pipes/number_pipe_spec.ts [hunk 1]
```diff
@@ -76,8 +76,14 @@ export function main() {
           expect(normalize(pipe.transform(5.1234, 'USD', false, '.0-3'))).toEqual('USD5.123');
         });
 
+        it('should not throw for undefined currencyCode',
+           () => expect(() => pipe.transform(12, undefined)).not.toThrowError());
```
Comment (pkozlowski-opensource) [inline modules/@angular/common/test/pipes/number_pipe_spec.ts]: This test is too weak - we should actually assert what is the valid result (in this case error thrown from the pipe would fail the test anyway so there is little value in testing just `.not.toThrowError()`).

File: modules/@angular/common/test/pipes/number_pipe_spec.ts [hunk 2]
```diff
@@ -76,8 +76,14 @@ export function main() {
...
+
+        it('should not throw for null currencyCode',
+           () => expect(() => pipe.transform(12, null)).not.toThrowError());
```
Comment (pkozlowski-opensource) [inline modules/@angular/common/test/pipes/number_pipe_spec.ts]: This test is too weak - we should actually assert what is the valid result (in this case error thrown from the pipe would fail the test anyway so there is little value in testing just `.not.toThrowError()`).

**Comment by pkozlowski-opensource** [comment]
@IgorMinar here is my thinking:

tl;dr; currently PR doesn't address the original issue (#7917) and the #7917 case is debatable. Also this PR would need more explicit tests. We shouldn't merge it as-is.

Details:

1. Currency pipe is allowing arbitrary changes in the currency code as of today: http://plnkr.co/edit/0hoqXGu70teUlmKgoAZw?p=preview. And hey, you can even skip the currency code and get dollars :-) http://plnkr.co/edit/FR3rUfHtQHh5Gdf0idFJ?p=preview. So in this sense this PR isn't making things worse - the problem is in place already. It just makes `{{100 | currency}}` and `{{100 | currency:null}}` equivalent.

2. Tests in this PR are not showing / testing what should actually happen when `null` is passed so this is why I've sticked the "cleanup" label on it. 

3. The way I'm reading the original issue (#7917) is that the intention was to display `100` for `{{100:null}}` instead of `USD100`. In this sense the PR doesn't address the opener issue.

4. One could argue if the original issue (#7917) makes sense. Personally I would say it is strange to see a number that all of the sudden gets the currency code pre-appended later on. This sounds like a problem in the data flow design in an app. I will comment in the original issue.

5. IMO the currency pipe is broken as-is since IMO it shouldn't default to any currency code but rather explicitly require a valid currency code. I would vote for removal of the default (USD) currency code and require a valid one.

@DzmitryShylovich does the above make sense?

**Comment by DzmitryShylovich** [comment]
> The way I'm reading the original issue (#7917) is that the intention was to display 100 for {{100:null}} instead of USD100. In this sense the PR doesn't address the opener issue.

yeah, you're 100% right. I missed that part completely 😄 
I'll add better tests.

File: modules/@angular/common/src/pipes/number_pipe.ts [hunk 3]
```diff
@@ -7,13 +7,12 @@
...
 
 import {Inject, LOCALE_ID, Pipe, PipeTransform, Type} from '@angular/core';
-
 import {NumberFormatStyle, NumberFormatter} from '../facade/intl';
 import {NumberWrapper, isBlank, isPresent} from '../facade/lang';
-
 import {InvalidPipeArgumentError} from './invalid_pipe_argument_error';
 
 const _NUMBER_FORMAT_REGEXP = /^(\d+)?\.((\d+)(-(\d+))?)?$/;
+const DEFAULT_CODE: string = 'USD';
```
Comment (laskoviymishka) [inline modules/@angular/common/src/pipes/number_pipe.ts]: Maybe it worth to add this thing as `OpaqueToken` to make it injectable, and by default let it be '' (empty string) to fulfill initial issue requirement.

Comment (pkozlowski-opensource) [inline modules/@angular/common/src/pipes/number_pipe.ts]: Let's not do that in this PR, this is a separate issue (and would be a good way of handling default currency)

**Comment by laskoviymishka** [comment]
@pkozlowski-opensource to add more flexibility maybe worth to create this default staff as `OpaqueToken`. In this case let it be defaulted to '' (empty string) so initial issue would ok, and some users would have an ability to override it to anything else.

**Comment by pkozlowski-opensource** [comment]
> I'll add better tests.

@DzmitryShylovich hold your horses, let me discus this with @IgorMinar and @vicb . IMO `{{100:currency}}` / `{{100:currency:null}}` should behave the same and should not be allowed (unless we want to have a notation of "default currency code" which is not a bad idea).

In any case we should _not_ have `{{100 | currency}}` to display `USD100` and `{{100 | currency:null}}` to display `100` - IMO both need to be consistent.

**Comment by pkozlowski-opensource** [comment]
> to add more flexibility maybe worth to create this default staff as OpaqueToken. In this case let it be defaulted to '' (empty string) so initial issue would ok, and some users would have an ability to override it to anything else.

@laskoviymishka yes, why not. Could you please open a separate issue for this? Let's not mix everything in this PR (there is enough discussion going on already).

File: modules/@angular/common/src/pipes/number_pipe.ts (hunk 3 above)
Comment (vicb) [inline modules/@angular/common/src/pipes/number_pipe.ts]: To me use case is that you want to set the number of digits without overriding the region default currency.

However IIRC Intl does not have default currency per region an this should come from a service we need to create (similar the the service returning the plural rules, see scripts/cldr)

**Comment by pkozlowski-opensource** [comment]
@DzmitryShylovich Thnx for the PR but I'm going to close this one for now given that:
* `null` / `undefined` should have the same effect as providing no currency code 
* we are having hard time deciding what the "default" currency means. 

I can see roughly 2 options here:
* having a configurable default
* having a service that gives you default based on regional settings (although this is a very dangerous path since we just can't take an arbitrary number and add a currency code to it!)
